From 408ab8fcc394f963a1d5750ecfc8906bd8990c07 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Rymanowski?= <lukasz.rymanowski@codecoup.pl>
Date: Wed, 6 May 2020 13:19:43 +0000
Subject: [PATCH] bluetooth/smp: Fix BT Security mode 1 level 4

As per Bluetooth core specification 5.2, Vol 3, Part C, 10.2.1

"Authenticated LE Secure Connections pairing with encryption using a 128-
bit strength encryption key."
---
 net/bluetooth/smp.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/net/bluetooth/smp.c b/net/bluetooth/smp.c
index 1476a91ce935..7a8c0904e4d9 100644
--- a/net/bluetooth/smp.c
+++ b/net/bluetooth/smp.c
@@ -730,6 +730,11 @@ static u8 check_enc_key_size(struct l2cap_conn *conn, __u8 max_key_size)
 	struct hci_dev *hdev = conn->hcon->hdev;
 	struct smp_chan *smp = chan->data;

+	/* Bluetooth core specification 5.2, Vol 3, Part C, 10.2.1 */
+	if ((conn->hcon->pending_sec_level == BT_SECURITY_FIPS)
+			&& (max_key_size != SMP_MAX_ENC_KEY_SIZE))
+		return SMP_ENC_KEY_SIZE;
+
 	if (max_key_size > hdev->le_max_key_size ||
 	    max_key_size < SMP_MIN_ENC_KEY_SIZE)
 		return SMP_ENC_KEY_SIZE;
--
2.20.1
